File: java/src/main/java/KMeans.java

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Iterator;

public class KMeans {

	private int n = 10;
	private int iters = 15;

	private Point[] Xs;
	private Point[] Centroids;

	public KMeans(Point[] Xs, int n, int iters) {
		this.Xs = Xs;
		this.n = n;
		this.iters = iters;
	}

	public KMeans(Point[] Xs) {
		this.Xs = Xs;
	}

	private double dist(Point p1, Point p2) {
		return p1.sub(p2).modulus();
	}

	private Point average(ArrayList<Point> xs) {
		double size = xs.size();
		Iterator<Point> iter = xs.iterator();
		Point p = new Point(0,0);
		while (iter.hasNext()) {
			p.addToThis(iter.next());
		}
		p.divideToThis(size);
		return p;
	}

	private Point closest(Point x, Point[] choices) {
		double minVal = dist(x, choices[0]);
		int min = 0;
		for (int i=1;i<choices.length;i++) {
			double actualDist = dist(x, choices[i]);
			if (actualDist < minVal) {
				min = i;
				minVal = actualDist;
			}
		}
		return choices[min];
	}

	private Collection<ArrayList<Point>> clusters() {
		HashMap<Point, ArrayList<Point>> hm = new HashMap<Point, ArrayList<Point>>();
		for (int i=0;i<Xs.length; i++) {
			Point Key = closest(Xs[i], Centroids);
			ArrayList<Point> alp = hm.get(Key);
			if (alp == null)
				alp = new ArrayList<Point>();
			alp.add(Xs[i]);
			hm.put(Key, alp);
		}
		return hm.values();
	}

	public Collection<ArrayList<Point>> run() {
		Centroids = new Point[n];
		for(int i=0;i<n;i++)
			Centroids[i] = Xs[i];

		for(int i=0;i<iters;i++) {
			Collection<ArrayList<Point>> clusters = clusters();
			int k = 0;
			for (ArrayList<Point> cluster: clusters) {
				Centroids[k] = average(cluster);
				k++;
			}
		}
		return clusters();
	}

}

File: swift/kmeans.swift
import Darwin

struct Point : Equatable {
  let x: Double
  let y: Double


  func modulus() -> Double {
    return sqrt(sq(self.x) + sq(self.y))
  }
}

extension Point: Hashable {
  var hashValue: Int {
      return x.hashValue ^ (y.hashValue << 1)
  }
}

func ==(x: Point, y: Point) -> Bool {
  return (x.x == y.x && x.y == y.y)
}

func +(left: Point, right: Point) -> Point {
  return Point(x: left.x + right.x, y: left.y + right.y)
}

func -(left: Point, right: Point) -> Point {
  return Point(x: left.x - right.x, y: left.y - right.y)
}

func /(left: Point, right: Double) -> Point {
  return Point(x: left.x / right, y: left.y / right)
}

func sq(x: Double) -> Double {
  return x*x
}

func dist(p1: Point, p2: Point) -> Double {
  return ((p1 - p2).modulus())
}

func average(xs: [Point]) -> Point {
  return xs.reduce(Point(x: 0, y: 0), combine: +) / Double(xs.count)
}

func closest(x: Point, choices: [Point]) -> Point {
  func comp(px: (point: Point, distance: Double), py: (point: Point, distance: Double)) -> (point: Point, distance: Double) {
    if (px.distance < py.distance) { return px }
    else { return py }
  }
  let minp = choices.map(
    {(point: $0, distance: (dist($0, p2: x)))}).reduce(
    (point: Point(x: 0, y: 0), distance: DBL_MAX), combine: comp)
  
  return minp.point
}

func clusters(xs: [Point], centroids: [Point]) -> [[Point]] {
  var dict =  [Point: [Point]]()

  for p in xs {
    let close = closest(p, choices: centroids)

    dict[close]?.append(p) ?? {dict[close] = [p]}()
  }
  
  return [[Point]](dict.values)
}

func run(xs: [Point], n: Int, iters: Int) -> [[Point]] {
  var centroids = [Point](xs[0..<n])

  for _ in 0..<iters {
    centroids = clusters(xs, centroids: centroids).map(average)
  }
/*
  print("Final centroids?")
  for c in centroids {
    print(c)
  }
*/    
  return clusters(xs, centroids: centroids)
}
